File: 82 Stars/javascript/stars.js
// STARS
//
// Converted from BASIC to Javascript by Qursch
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var guesses = 7;
var limit = 100;

// Main program
async function main()
{
    print(tab(33) + "STARS\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n\n\n");

    print("DO YOU WANT INSTRUCTIONS? (Y/N)");
    var instructions = await input();
    if(instructions.toLowerCase()[0] == "y") {
        print(`I AM THINKING OF A WHOLE NUMBER FROM 1 TO ${limit}\n`);
        print("TRY TO GUESS MY NUMBER.  AFTER YOU GUESS, I\n");
        print("WILL TYPE ONE OR MORE STARS (*).  THE MORE\n");
        print("STARS I TYPE, THE CLOSER YOU ARE TO MY NUMBER.\n");
        print("ONE STAR (*) MEANS FAR AWAY, SEVEN STARS (*******)\n");
        print(`MEANS REALLY CLOSE!  YOU GET ${guesses} GUESSES.\n\n\n`);
    }
    while (true) {
        var randomNum = Math.floor(Math.random() * 150) + 1;
        console.log(randomNum);

        print("OK, I AM THINKING OF A NUMBER, START GUESSING.\n\n");

        var loss = true;

        for(var guessNum=1; guessNum <= guesses; guessNum++) {
            print("YOUR GUESS");

            var guess = parseInt(await input());

            if(guess == randomNum) {
                loss = false;
                print("*".repeat(79) + "!!!\n");
                print(`YOU GOT IT IN ${guessNum} GUESSES!!! LET'S PLAY AGAIN...\n\n`);
                break;
            }

            var dist = Math.abs(guess - randomNum);
              
            //IMPLEMENT CHECK HERE
        }

        if(loss) {
            print(`SORRY, THAT'S ${guesses} GUESSES. THE NUMBER WAS ${randomNum}\n\n`);
        }
    }
}

main();

File: 82 Stars/java/src/Stars.java
import java.util.Arrays;
import java.util.Scanner;

/**
 * Game of Stars
 *
 * Based on the Basic game of Stars here
 * https://github.com/coding-horror/basic-computer-games/blob/main/82%20Stars/stars.bas
 *
 * Note:  The idea was to create a version of the 1970's Basic game in Java, without introducing
 *        new features - no additional text, error checking, etc has been added.
 */
public class Stars {

    public static final int HIGH_NUMBER_RANGE = 100;
    public static final int MAX_GUESSES = 7;

    private enum GAME_STATE {
        STARTING,
        INSTRUCTIONS,
        START_GAME,
        GUESSING,
        WON,
        LOST,
        GAME_OVER
    }

    // Used for keyboard input
    private final Scanner kbScanner;

    // Current game state
    private GAME_STATE gameState;

    // Players guess count;
    private int playerTotalGuesses;

    // Players current guess
    private int playerCurrentGuess;

    // Computers random number
    private int computersNumber;

    public Stars() {

        gameState = GAME_STATE.STARTING;

        // Initialise kb scanner
        kbScanner = new Scanner(System.in);
    }

    /**
     * Main game loop
     *
     */
    public void play() {

        do {
            switch (gameState) {

                // Show an introduction the first time the game is played.
                case STARTING:
                    intro();
                    gameState = GAME_STATE.INSTRUCTIONS;
                    break;

                // Ask if instructions are needed and display if yes
                case INSTRUCTIONS:
                    if(yesEntered(displayTextAndGetInput("DO YOU WANT INSTRUCTIONS? "))) {
                        instructions();
                    }
                    gameState = GAME_STATE.START_GAME;
                    break;

                // Generate computers number for player to guess, etc.
                case START_GAME:
                    init();
                    System.out.println("OK, I AM THINKING OF A NUMBER, START GUESSING.");
                    gameState = GAME_STATE.GUESSING;
                    break;

                // Player guesses the number until they get it or run out of guesses
                case GUESSING:
                    playerCurrentGuess = playerGuess();

                    // Check if the player guessed the number
                    if(playerCurrentGuess == computersNumber) {
                        gameState = GAME_STATE.WON;
                    } else {
                        // incorrect guess
                        showStars();
                        playerTotalGuesses++;
                        // Ran out of guesses?
                        if (playerTotalGuesses > MAX_GUESSES) {
                            gameState = GAME_STATE.LOST;
                        }
                    }
                    break;

                // Won game.
                case WON:

                    System.out.println(stars(79));
                    System.out.println("YOU GOT IT IN " + playerTotalGuesses
                            + " GUESSES!!!  LET'S PLAY AGAIN...");
                    gameState = GAME_STATE.START_GAME;
                    break;

                // Lost game by running out of guesses
                case LOST:
                    System.out.println("SORRY, THAT'S " + MAX_GUESSES
                            + " GUESSES. THE NUMBER WAS " + computersNumber);
                    gameState = GAME_STATE.START_GAME;
                    break;
            }
            // Endless loop since the original code did not allow the player to exit
        } while (gameState != GAME_STATE.GAME_OVER);
    }

    /**
     * Shows how close a players guess is to the computers number by
     * showing a series of stars - the more stars the closer to the
     * number.
     *
     */
    private void showStars() {
        int d = Math.abs(playerCurrentGuess - computersNumber);
        int starsToShow;
        if(d >=64) {
            starsToShow = 1;
        } else if(d >=32) {
            starsToShow = 2;
        } else if (d >= 16) {
            starsToShow = 3;
        } else if (d >=8) {
            starsToShow = 4;
        } else if( d>= 4) {
            starsToShow = 5;
        } else if(d>= 2) {
            starsToShow = 6;
        } else {
            starsToShow = 7;
        }
        System.out.println(stars(starsToShow));
    }

    /**
     * Show a number of stars (asterisks)
     * @param number the number of stars needed
     * @return the string encoded with the number of required stars
     */
    private String stars(int number) {
        char[] stars = new char[number];
        Arrays.fill(stars, '*');
        return new String(stars);
    }

    /**
     * Initialise variables before each new game
     *
     */
    private void init() {
        playerTotalGuesses = 1;
        computersNumber = randomNumber();
    }

    public void instructions() {
        System.out.println("I AM THINKING OF A WHOLE NUMBER FROM 1 TO " + HIGH_NUMBER_RANGE);
        System.out.println("TRY TO GUESS MY NUMBER.  AFTER YOU GUESS, I");
        System.out.println("WILL TYPE ONE OR MORE STARS (*).  THE MORE");
        System.out.println("STARS I TYPE, THE CLOSER YOU ARE TO MY NUMBER.");
        System.out.println("ONE STAR (*) MEANS FAR AWAY, SEVEN STARS (*******)");
        System.out.println("MEANS REALLY CLOSE!  YOU GET " + MAX_GUESSES + " GUESSES.");
    }

    public void intro() {
        System.out.println("STARS");
        System.out.println("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
        System.out.println();
    }

    /**
     * Get players guess from kb
     *
     * @return players guess as an int
     */
    private int playerGuess() {
        return Integer.parseInt((displayTextAndGetInput("YOUR GUESS? ")));
    }

    /**
     * Checks whether player entered Y or YES to a question.
     *
     * @param text  player string from kb
     * @return true of Y or YES was entered, otherwise false
     */
    private boolean yesEntered(String text) {
        return stringIsAnyValue(text, "Y", "YES");
    }

    /**
     * Check whether a string equals one of a variable number of values
     * Useful to check for Y or YES for example
     * Comparison is case insensitive.
     *
     * @param text source string
     * @param values a range of values to compare against the source string
     * @return true if a comparison was found in one of the variable number of strings passed
     */
    private boolean stringIsAnyValue(String text, String... values) {

        // Cycle through the variable number of values and test each
        for(String val:values) {
            if(text.equalsIgnoreCase(val)) {
                return true;
            }
        }

        // no matches
        return false;
    }

    /*
     * Print a message on the screen, then accept input from Keyboard.
     *
     * @param text message to be displayed on screen.
     * @return what was typed by the player.
     */
    private String displayTextAndGetInput(String text) {
        System.out.print(text);
        return kbScanner.next();
    }

    /**
     * Generate random number
     *
     * @return random number
     */
    private int randomNumber() {
        return (int) (Math.random()
                * (HIGH_NUMBER_RANGE) + 1);
    }
}